provisioning/device/src/ProvisioningDeviceClient.cs

```diff
@@ -79,12 +80,63 @@ public Task<DeviceRegistrationResult> RegisterAsync()
         /// <param name="cancellationToken">The cancellation token.</param>
         /// <returns>The registration result.</returns>
         public Task<DeviceRegistrationResult> RegisterAsync(CancellationToken cancellationToken)
+        {
+            return RegisterAsync(default(RegisterInput), cancellationToken);
+        }
+
+        /// <summary>
+        /// Registers the current device using the Device Provisioning Service and assigns it to an IoT Hub.
+        /// </summary>
+        /// <param name="registerInput">The register input.</param>
+        /// <returns>The registration result.</returns>
+        public Task<DeviceRegistrationResult> RegisterAsync(RegisterInput registerInput)
+        {
+            return RegisterAsync(registerInput, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Registers the current device using the Device Provisioning Service and assigns it to an IoT Hub.
+        /// </summary>
+        /// <param name="registerInput">The register input.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The registration result.</returns>
+        public Task<DeviceRegistrationResult> RegisterAsync(RegisterInput registerInput, CancellationToken cancellationToken)
         {
             if (Logging.IsEnabled) Logging.RegisterAsync(this, _globalDeviceEndpoint, _idScope, _transport, _security);
 
-            var request = new ProvisioningTransportRegisterMessage(_globalDeviceEndpoint, _idScope, _security);
+            var request = new ProvisioningTransportRegisterMessage(_globalDeviceEndpoint, _idScope, _security, registerInput?.Data);
             request.ProductInfo = ProductInfo;
-            return _transport.RegisterAsync(request, cancellationToken);
+
+            ExponentialBackoff retryPolicy = new ExponentialBackoff();
+            int retryCount = 0;
+
```

Comment:
Is this expected? I was under the impression that we are not implementing retry policies for DPS at the moment and we decided to retry at transport level for certain failures (I think beyond 5xx)

Reply:
I'll look into what Node and C do for this retry

Reply:
@rajeevmv if DPS returns a 429 or 500 to an operation status lookup call, can I expect that it will still parse into a RegistrationOperationStatus object? And since these are transient errors, can I still expect that parsed RegistrationOperationStatus object to have the status of "Assigning" still?

Reply:
No. The 429/400 response will not contain a parseable response object other than the error code and error details.

---
In reply to: [247653140](https://github.com/Azure/azure-iot-sdk-csharp/pull/750#discussion_r247653140) [](ancestors = 247653140)

Reply:
I'm going to separate the reprovisioning (and its retry) into a separate PR for clarity. This PR will only contain the support for Symmetric Key attestation.